servicetalk-dns-discovery-netty/src/test/java/io/servicetalk/dns/discovery/netty/DefaultDnsClientTest.java

```diff
@@ -914,11 +874,127 @@ private static void assertHasEvent(Collection<ServiceDiscovererEvent<InetAddress
+
+        private final EventLoopAwareNettyIoExecutor delegate;
+        private final TestExecutor timer;
+
+        NettyIoExecutorWithTestTimer(EventLoopAwareNettyIoExecutor delegate, TestExecutor timer) {
+            this.delegate = delegate.next();    // Take a single EventLoopIoExecutor
+            this.timer = timer;
+        }
+
+        @Override
+        public Completable closeAsync() {
+            return delegate.closeAsync();
+        }
+
+        @Override
+        public Completable closeAsyncGracefully() {
+            return delegate.closeAsyncGracefully();
+        }
+
+        @Override
+        public Completable onClose() {
+            return delegate.onClose();
+        }
+
+        @Override
+        public Completable onClosing() {
+            return delegate.onClosing();
+        }
+
+        @Override
+        public boolean isUnixDomainSocketSupported() {
+            return delegate.isUnixDomainSocketSupported();
+        }
+
+        @Override
+        public boolean isFileDescriptorSocketAddressSupported() {
+            return delegate.isFileDescriptorSocketAddressSupported();
+        }
+
+        @Override
+        public boolean isIoThreadSupported() {
+            return delegate.isIoThreadSupported();
+        }
+
+        @Override
+        public boolean isCurrentThreadEventLoop() {
+            return delegate.isCurrentThreadEventLoop();
+        }
+
+        @Override
+        public EventLoopGroup eventLoopGroup() {
+            return delegate.eventLoopGroup();
+        }
+
+        @Override
+        public EventLoopAwareNettyIoExecutor next() {
+            return this;
+        }
+
+        @Override
+        public Cancellable execute(final Runnable task) throws RejectedExecutionException {
+            return delegate.execute(task);
+        }
+
+        @Override
+        public Completable submit(final Runnable runnable) {
+            return delegate.submit(runnable);
+        }
+
+        @Override
+        public Completable submitRunnable(final Supplier<Runnable> runnableSupplier) {
+            return delegate.submitRunnable(runnableSupplier);
+        }
+
+        @Override
+        public <T> Single<T> submit(final Callable<? extends T> callable) {
+            return delegate.submit(callable);
+        }
+
+        @Override
+        public <T> Single<T> submitCallable(final Supplier<? extends Callable<? extends T>> callableSupplier) {
+            return delegate.submitCallable(callableSupplier);
+        }
+
+        @Override
+        public Executor asExecutor() {
+            return timer;
+        }
+
+        @Override
+        public long currentTime(final TimeUnit unit) {
+            return timer.currentTime(unit);
+        }
+
+        @Override
+        public Cancellable schedule(final Runnable task, final long delay, final TimeUnit unit)
+                throws RejectedExecutionException {
+            return timer.schedule(() -> delegate.execute(task), delay, unit);
```

Comment:
is that task run on the timer threads or the delegate threads?

Reply:
If we pass original `IoExecutor`, it will schedule and execute tasks on EventLoop. With this modification, we move scheduling to `TestExecutor`, then have to manually offload execution back to EventLoop, because internal implementation of `DefaultDnsClient` expects it with multiple assertions. I will leave a comment in code.